### ISSUE TITLE
Mysterious "Some packages have new versions but cannot be upgraded" message

### ISSUE THREAD

**Issue by fingolfin** [JuliaLang/Pkg.jl#2989]
I really like the new `status --outdated` feature coming in Julia 1.8! It relieves a serious pain point that's been bothering some of us for years. Thanks!

There is a minor weirdness about it, though: in some cases it keeps telling me
> Info Some packages have new versions but cannot be upgraded. To see why use `status --outdated`

But entering that command shows no packages:
```
$ julia-master
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.1558 (2022-02-15)
 _/ |\__'_|_|_|\__'_|  |  Commit 9542577c88* (0 days old master)
|__/                   |

(@v1.8) pkg> update
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
  No Changes to `~/.julia/environments/v1.8/Project.toml`
  No Changes to `~/.julia/environments/v1.8/Manifest.toml`
        Info Some packages have new versions but cannot be upgraded. To see why use `status --outdated`

(@v1.8) pkg> status --outdated
Status `~/.julia/environments/v1.8/Project.toml`

(@v1.8) pkg> status
Status `~/.julia/environments/v1.8/Project.toml`
  [6e4b80f9] BenchmarkTools v1.3.1
  [c3a54625] JET v0.5.6
  [f1435218] Oscar v0.8.1-DEV `~/.julia/dev/Oscar`
  [295af30f] Revise v3.3.1
  [aa65fe97] SnoopCompile v2.9.0
  [e2b509da] SnoopCompileCore v2.9.0

(@v1.8) pkg>
```

The answer, "of course" (well... as I realize while typing this, but after having been puzzled by it for several days) is that there *is* a package in my manifest that is outdated, but it is only a dependency of one of my packages:
```
(v1.8) pkg> status --outdated -m
Status `/tmpbig/horn/julia_depot/environments/v1.8/Manifest.toml`
⌅ [4fe8b98c] Mongoc v0.6.2 (<v0.7.1): Polymake
```

I see two solutions for this:

1. don't show that message about outdated packages in this situation, i.e., if it only affects things which dependencies of the packages I actually care about.

2. detect this situation, and then change the recommendation shown to the user, and suggest to enter `status --outdated --manifest`

Personally I prefer 1: I don't care about those indirect dependencies, I have to assume that the packages I use "know what they are doing".

**Comment by IanButterworth** [comment]
Yeah, 1 does sound better

**Comment by DilumAluthge** [comment]
Personally I disagree... I want to know about outdated indirect dependencies.

**Comment by KristofferC** [comment]
You do if you pass the `-m` flag to status though?

**Comment by DilumAluthge** [comment]
Sure, but it's nice to get the message even if I just do `Pkg.status()`.

I think option 2 is the best choice here.

**Comment by DilumAluthge** [comment]
I personally think that option 1 is a step backwards, and that option 2 would solve this issue just fine.

**Comment by DilumAluthge** [comment]
If an indirect dependency is being held back to an older version, this can definitely have an impact on the behavior and/or performance of the direct dependencies.

So I disagree with the idea that most users can ignore it if their indirect dependencies are being held back.

**Comment by IanButterworth** [comment]
Both options:
https://github.com/JuliaLang/Pkg.jl/pull/2990
https://github.com/JuliaLang/Pkg.jl/pull/2991

**Comment by IanButterworth** [comment]
I'm ok with either. But I get @DilumAluthge point. Perhaps let's go with https://github.com/JuliaLang/Pkg.jl/pull/2991 and see if it grinds people's gears

**Comment by KristofferC** [comment]
`status` has always shown information about direct dependencies, `status -m` shows it for manifest. This is completely consistent.

**Comment by IanButterworth** [comment]
Which do you mean by "this"?

To be clear, this isn't about `pkg> status` it's about the combined mode status print at the end of `pkg> update`.

If there were hidden outdated deps in both the project and manifest you'd get the warning twice, in each.

**Comment by IanButterworth** [comment]
i.e. option 2 could result in
```
(@v1.8) pkg> update
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
  No Changes to `~/.julia/environments/v1.8/Project.toml`
        Info Some packages have new versions but cannot be upgraded. To see why use `status --outdated`
  No Changes to `~/.julia/environments/v1.8/Manifest.toml`
        Info Some packages have new versions but cannot be upgraded. To see why use `status --outdated --manifest`
```

**Comment by DilumAluthge** [comment]
That seems perfectly fine to me. There are outdated direct dependencies, and I can use `status --outdated` to list them. There are also outdated indirect dependencies, and I can use `status --outdated --manifest` to list those.

The important part, in my opinion, is to make sure we show these warnings when people do `] update`, because we have seen plenty of users on Discourse that believe that `] update` will always update all packages, and are surprised when this does not happen.

The original problem reported in this issue was that `] update` told the user to do `status --outdated`, but then `status --outdated` did not show any packages. The only thing we need to do, then, is to tell the user to do `status --outdated --manifest`, which completely solves the problem.

**Comment by fingolfin** [comment]
First of: thank you all for immediately working on this, and esp. @IanButterworth for submitting *two* PRs to fix this in separate ways. Much appreciated. Either way will be an improvement!

As to the choice between the two possible changes, I think *for me* in the end it boils down to whether I have my "package developer hat" or my "user hat" on:

> If an indirect dependency is being held back to an older version, this can definitely have an impact on the behavior and/or performance of the direct dependencies.

As a package *user*, I think I'll find this very annoying: There is usually nothing I can do about it, can I? I specified the wanted/needed packages; all of them are already at the latest version. So that's fine for me! If this combination leads to some indirect dependency being outdated, then that's how it is. All should work. If not, then arguably it is a bug in one or more of the involved packages, which failed to correctly specify version restrictions on one of its dependencies. But what can I do about it? I know nothing about that outdated package. Is this a problem? I am getting anxious now 😓 

As a package *developer* of one of the involved packages: for debugging reports from users, it'd be so helpful to know that some dependency was being held back. Typical situation: user X is using my package A together with some package B I never heard of. Somehow their (possibly quite indirect) dependencies have an interaction that holds back C from being the latest... But C had that bug fix that was important to me. So if user X had reported this outdated package from the start, that would have helped me to resolve their bug report much quicker...

... except that in reality, my users never mention such things 😢 so I always end up asking for `status --manifest` anyway. So from that point of view, I really don't see the benefit of reporting the outdated indirect deps.


tl;dr: for *me* personally, on my computers, either solution will be fine, as I understand the technical underpinnings sufficiently well to be able to process it. But I know a lot of end users of our packages who don't, and I'll have to process their support requests -- so I guess my preference is rather selfish 😂 .

**Comment by KristofferC** [comment]
> The original problem reported in this issue was that ] update told the user to do status --outdated, but then status --outdated did not show any packages. The only thing we need to do, then, is to tell the user to do status --outdated --manifest, which completely solves the problem.

No, the problem was that it showed anything at all. 

> As a package user, I think I'll find this very annoying: There is usually nothing I can do about it, can I? 

Yes, exactly.

**Comment by IanButterworth** [comment]
I think I'm siding with option 1: https://github.com/JuliaLang/Pkg.jl/pull/2990

Error/warning fatigue is real, and from my experience on 1.8 so far the hidden manifest outdated message is shown all most all the time for real world environments. 

Lets give https://github.com/JuliaLang/Pkg.jl/pull/2990 a go for the first 1.8 cut

**Comment by ufechner7** [comment]
Still present in 1.8-rc1. Shall this be re-opened?

**Comment by IanButterworth** [comment]
@ufechner7 can you give an example? 
I don't see it in 1.8.0-rc1. See here which would've shown the message under the no-change manifest prior to this PR

```
(@v1.8) pkg> up
    Updating registry at `~/.julia/registries/General.toml`
  No Changes to `~/.julia/environments/v1.8/Project.toml`
  No Changes to `~/.julia/environments/v1.8/Manifest.toml`

(@v1.8) pkg> st --outdated -m
Status `~/.julia/environments/v1.8/Manifest.toml`
⌅ [8197267c] IntervalSets v0.6.2 (<v0.7.1): GtkObservables, ProfileView
```

**Comment by KristofferC** [comment]
Thera are two things.

When installing packages the message will be shown even if there are only packages in the manifest that are outdated:

```
(jl_Qkbsmu) pkg> add KiteControllers
   Resolving package versions...
    Updating `/tmp/jl_Qkbsmu/Project.toml`
  [b19fd437] + KiteControllers v0.1.0
    Updating `/tmp/jl_Qkbsmu/Manifest.toml`
  [79e6a3ab] + Adapt v3.3.3
...
  [69666777] + Arrow v2.3.0
⌅ [31f734f8] + ArrowTypes v1.2.1
  [c3b6d118] + BitIntegers v0.2.6
...
⌅ [34da2185] + Compat v3.44.0
  [9a962f9c] + DataAPI v1.10.0
  [e2d170a0] + DataValueInterfaces v1.0.0
⌅ [ffbed154] + DocStringExtensions v0.8.6
  [fa6b7ba4] + DualNumbers v0.6.8
...
⌅ [90980105] + KiteUtils v0.3.5
  [929cbde3] + LLVM v4.13.0
...
        Info Packages marked with ⌅ have new versions available but cannot be upgraded. To see why use `status --outdated`

(jl_Qkbsmu) pkg> status --outdated
Status `/tmp/jl_Qkbsmu/Project.toml`

(jl_Qkbsmu) pkg> status --outdated -m
Status `/tmp/jl_Qkbsmu/Manifest.toml`
⌅ [31f734f8] ArrowTypes v1.2.1 (<v2.0.0): Arrow
⌅ [34da2185] Compat v3.44.0 (<v4.1.0): FilePathsBase, Mocking
⌅ [ffbed154] DocStringExtensions v0.8.6 (<v0.9.0): KiteUtils, LogExpFunctions, RecursiveArrayTools
⌅ [90980105] KiteUtils v0.3.5 (<v0.4.7): KiteControllers
```

And when running the tests, it isn't helpful to print "To see why use `status --outdated`" because you are in a temporary project.

**Comment by IanButterworth** [comment]
But I would propose opening issues for those and not reopening this one. The original issue here is fixed, as shown in my example.

**Comment by KristofferC** [comment]
> Not sure how to handle this. Perhaps use isinteractive()?

Just pass a flag to `status` that tells it not to show that part I guess is the easiest. Here I  think: https://github.com/JuliaLang/Pkg.jl/blob/201cbe2a50c06db65b1e7a30f64d4875042f2d4d/src/Operations.jl#L1744
